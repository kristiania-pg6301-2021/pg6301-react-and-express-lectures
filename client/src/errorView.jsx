import React from "react";

export function ErrorView({ error, reload }) {
  return (
    <div>
      <h1>An error occurred</h1>
      <div>{error.toString()}</div>
      <button onClick={reload}>Retry</button>
    </div>
  );
}