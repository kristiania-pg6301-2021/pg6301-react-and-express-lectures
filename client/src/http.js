class HttpError extends Error {
  constructor(url, status, statusText) {
    super(`${url}: ${status} ${statusText}`);
  }
}

export async function fetchJSON(url, options) {
  const response = await fetch(url, options);
  if (!response.ok) {
    throw new HttpError(url, response.status, response.statusText);
  }
  return await response.json();
}
