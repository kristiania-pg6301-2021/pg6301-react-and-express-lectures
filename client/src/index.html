<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <title>KristianiaQuiz</title>
  </head>
  <body>
    <div id="app"></div>
  </body>
  <script src="index.jsx"></script>
</html>
