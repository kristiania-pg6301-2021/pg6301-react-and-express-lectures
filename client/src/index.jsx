import React from "react";
import ReactDOM from "react-dom";
import { BrowserRouter, Link } from "react-router-dom";
import { Route, Switch } from "react-router";

function Application() {
  return (
    <BrowserRouter>
      <header>
        <Link to={"/"}>KristianiaQuiz</Link>
      </header>
      <main>
        <Switch>
          <Route path="/login">
            <h1>Login</h1>
          </Route>
          <Route path="/" exact>
            <h1>Home page</h1>
            <ul>
              <li>
                <Link to={"/login"}>Login</Link>
              </li>
            </ul>
          </Route>
        </Switch>
      </main>
    </BrowserRouter>
  );
}

ReactDOM.render(<Application />, document.getElementById("app"));
