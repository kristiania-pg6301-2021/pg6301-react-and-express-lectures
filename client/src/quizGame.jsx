import React from "react";
import { useLoader } from "./useLoader";
import { ErrorView } from "./errorView";
import { LoadingView } from "./loadingView";

export function QuizGame({ api }) {
  const { data, loading, error, reload } = useLoader(api.fetchQuiz);

  async function startGame() {
    await api.startQuiz();
    reload();
  }

  if (error) {
    return <ErrorView error={error} reload={reload} />;
  }
  if (loading) {
    return <LoadingView />;
  }

  if (data.state === "not_started") {
    return (
      <div>
        <div>Do you want to play a game?</div>
        <button onClick={startGame}>Start game</button>
      </div>
    );
  }

  async function handleAnswer(index) {
    await api.answerQuestion(index);
    reload();
  }

  const { question, alternatives } = data;

  return (
    <div>
      <h2>{question}</h2>
      {alternatives.map((a, index) => (
        <button key={index} onClick={() => handleAnswer(index)}>
          {a}
        </button>
      ))}
    </div>
  );
}
