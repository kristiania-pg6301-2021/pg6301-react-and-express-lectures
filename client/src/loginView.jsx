import React, { useEffect, useState } from "react";
import { fetchJSON } from "./http";
import { ErrorView } from "./errorView";
import { LoadingView } from "./loadingView";

export function LoginView({ loginProvider, onLoggedIn }) {
  const redirect_uri = window.location.origin;
  const hash = Object.fromEntries(
    new URLSearchParams(window.location.hash.substr(1))
  );
  const [error, setError] = useState();
  const [loading, setLoading] = useState(true);

  async function authorize() {
    const { discovery_url, client_id, scope } = loginProvider;
    const { authorization_endpoint } = await fetchJSON(discovery_url);

    const loginState = {
      state: randomString(30),
      code_verifier: randomString(50),
    };
    sessionStorage.setItem("loginState", JSON.stringify(loginState));
    const { state, code_verifier } = loginState;

    const params = {
      client_id,
      redirect_uri,
      response_type: "code",
      response_mode: "fragment",
      scope,
      state,
      code_challenge_method: "S256",
      code_challenge: await sha256(code_verifier),
    };

    window.location.href =
      authorization_endpoint + "?" + new URLSearchParams(params);
  }

  async function fetchToken() {
    const { code, error } = hash;

    if (error) {
      return setError(error);
    }
    if (!code) {
      setLoading(false);
      return;
    }

    try {
      const { discovery_url, client_id } = loginProvider;
      const { token_endpoint } = await fetchJSON(discovery_url);
      const loginState = JSON.parse(sessionStorage.getItem("loginState"));
      const { code_verifier } = loginState;
      const payload = {
        client_id,
        code,
        redirect_uri,
        code_verifier,
        grant_type: "authorization_code",
      };

      const tokenResponse = await fetchJSON(token_endpoint, {
        method: "POST",
        body: new URLSearchParams(payload),
      });
      localStorage.setItem("authorization", JSON.stringify(tokenResponse));
      sessionStorage.removeItem("loginState");
      onLoggedIn();
    } catch (e) {
      console.log("failed to get token");
      setError(e);
    } finally {
      console.log("done with token");
      setLoading(false);
      window.location.hash = "";
    }
  }
  useEffect(fetchToken, [hash]);

  if (error) {
    return <ErrorView error={error} reload={authorize} />;
  }
  if (loading) {
    return <LoadingView />;
  }

  return <button onClick={authorize}>Log me in</button>;
}

export function randomString(length) {
  const possible =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmopqrstuvwxyz1234567890";
  let result = "";
  for (let i = 0; i < length; i++) {
    result += possible.charAt(Math.floor(Math.random() * possible.length));
  }
  return result;
}

export async function sha256(string) {
  const binaryHash = await crypto.subtle.digest(
    "SHA-256",
    new TextEncoder("utf-8").encode(string)
  );
  return btoa(String.fromCharCode.apply(null, new Uint8Array(binaryHash)))
    .split("=")[0]
    .replace(/\+/g, "-")
    .replace(/\//g, "_");
}
