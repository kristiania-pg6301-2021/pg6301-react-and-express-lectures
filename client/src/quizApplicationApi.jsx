import { fetchJSON } from "./http";

export class QuizApplicationApi {
  async getUserinfo() {
    const { access_token } =
      JSON.parse(localStorage.getItem("authorization")) || {};
    return await fetchJSON("https://webapps.kristiania.no:3000/api/userinfo", {
      headers: {
        ...(access_token ? { Authorization: `Bearer ${access_token}` } : {}),
      },
    });
  }
}
