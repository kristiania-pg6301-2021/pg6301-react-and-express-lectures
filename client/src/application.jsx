import React from "react";
import { LoginView } from "./loginView";
import { ErrorView } from "./errorView";
import { LoadingView } from "./loadingView";
import { useLoader } from "./useLoader";
import { QuizGamePage } from "./quizGamePage";

export function Application({ api }) {
  return (
    <div>
      <h1>Kristiania Quiz</h1>
      <QuizGamePage api={api} />
    </div>
  );
}
