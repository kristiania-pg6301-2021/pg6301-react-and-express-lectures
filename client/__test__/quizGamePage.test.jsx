import { QuizGamePage } from "../src/quizGamePage";
import { MemoryRouter } from "react-router";
import * as ReactDOM from "react-dom";
import * as React from "react";
import { act } from "react-dom/test-utils";

describe("the quiz game page", () => {
  it("can display a starting page", async () => {
    const container = document.createElement("div");
    const fetchQuiz = () => ({
      state: "not_started",
    });
    await act(async () => {
      await ReactDOM.render(
        <MemoryRouter>
          <QuizGamePage api={{ fetchQuiz }} />
        </MemoryRouter>,
        container
      );
    });
    expect(container.innerHTML).toMatchSnapshot();
    expect(container.querySelector("button").textContent).toEqual("Start game");
  });
});
