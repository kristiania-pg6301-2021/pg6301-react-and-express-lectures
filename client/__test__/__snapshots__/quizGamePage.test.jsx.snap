// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the quiz game page can display a question 1`] = `"<div>Loading</div>"`;

exports[`the quiz game page can display a starting page 1`] = `"<div><div>Do you want to play a game?</div><button>Start game</button></div>"`;
